keen/tests/saved_query_tests.py

```diff
@@ -93,33 +97,106 @@ def test_create_saved_query(self):
 
     @responses.activate
     def test_update_saved_query(self):
-        saved_queries_response = {
+        unacceptable_attr = "run_information"
+        metadata_attr_name = "metadata"
+
+        original_query = {
             "query_name": "saved-query-name",
             "refresh_rate": 14400,
             "query": {
                 "analysis_type": "average",
                 "event_collection": "TheCollection",
                 "target_property": "TheProperty",
                 "timeframe": "this_2_weeks"
-                }
+            },
+            metadata_attr_name: { "foo": "bar" },
+            unacceptable_attr: { "foo": "bar" }
         }
+
         url = "{0}/{1}/projects/{2}/queries/saved/saved-query-name".format(
             self.client.api.base_url,
             self.client.api.api_version,
             self.exp_project_id
         )
 
         responses.add(
-            responses.GET, url, status=200, json=saved_queries_response
+            responses.GET, url, status=200, json=original_query
         )
 
-        responses.add(
-            responses.PUT, url, status=200, json=saved_queries_response
+        updated_query = { "query": {} }
+        new_analysis_type = "sum"
+        updated_query["query"]["analysis_type"] = new_analysis_type
+
+        def request_callback(request):
+            payload = json.loads(request.body)
+
+            # Ensure update() round-trips some necessary things like "metadata"
+            self.assertEqual(payload[metadata_attr_name], original_query[metadata_attr_name])
+
+            # Ensure update() doesn't pass unacceptable attributes
+            self.assertNotIn(unacceptable_attr, payload)
+
+            # Ensure update() merges deep updates
+            self.assertEqual(payload["query"]["analysis_type"], new_analysis_type)
+            payload["query"]["analysis_type"] = "average"
+            payload[unacceptable_attr] = original_query[unacceptable_attr]
+            self.assertEqual(payload, original_query)
+
+            headers = {}
+            return (200, headers, json.dumps(updated_query))
+
+        responses.add_callback(
+            responses.PUT,
+            url,
+            callback=request_callback,
+            content_type='application/json',
         )
 
-        saved_query = self.client.saved_queries.update("saved-query-name", saved_queries_response)
+        saved_query = self.client.saved_queries.update("saved-query-name", updated_query)
 
-        self.assertEquals(saved_query, saved_queries_response)
+        self.assertEqual(saved_query, updated_query)
```

Comment:
Seems odd to me to test that `saved_query` is equal to `update_query` here since `updated_query` isn't actually the complete query. I see that `updated_query` is returned from the request callback above, but why not use the full updated query after it has been validated? I'm guessing it's probably just inconvenient? I can see that as-is this does validate the logic that whatever gets returned from `update` is equal to the response body, which I suppose is the important part.